# Регулярные выражения -- продолжение

официальная документация модуля [re](https://docs.python.org/3/library/re.html)

Один из многих сайтов для отладки и демонстрации регулярных выражений <https://regex101.com/r/jTIHEj/1>

* * *

## С примерами  можно поиграться через визуальный интерфейс сайта regex101.com


<https://regex101.com/r/Gx7Dcz/1> конец строки

<https://regex101.com/r/rzd9XB/1> слово целиком

<https://regex101.com/r/jUrQBO/1> классы символов


<https://regex101.com/r/EtwAKb/1> цепочка гласных

<https://regex101.com/r/FbrUSV/1> звездочка

<https://regex101.com/r/qLfMh4/1>

<https://regex101.com/r/eX1dSY/1>  число после необязательных пробелов * +

<https://regex101.com/r/ATNMJG/1> альтернатива


<https://regex101.com/r/p9IkeS/1> greedy \* 

<https://regex101.com/r/BgeiQ7/1>

<https://regex101.com/r/V5o3Ei/1> non-greedy quantifiers \*? 

<https://regex101.com/r/RDCq0x/1>


#### группы 

<https://regex101.com/r/QIrvlI/1> 
<https://regex101.com/r/56hy6b/1>  

пример  применения нежадных квантификаторов `*?` и групп для подготовки файла к импорту. В файле проблемма: символ разделитель полей встречается в данных, нужно поле  в котором есть такая проблема окружить так: "Korea, Rep", KOR

Сценарий 1. В первом поле текстового csv-файда название страны, содержащее запятую, являющуюся разделителем в этом файле. Нужно заменой окружить это поле в двойные кавычки.

<https://regex101.com/r/jtL2Av/1>


Сценарий 2.  В первом поле текстового csv-файда название страны, не обязательно  содержащее запятую или несколько запятых, являющуюся разделителем в этом файле. Следующий столбец -- это код страны, состоящий из 3 больших букв. Нужно заменой окружить первое поле с названием страны  в двойные кавычки.
<https://regex101.com/r/JfpXXA/1>


##### look ahead/behind 

(когда нужно взять при помощи шаблона текст рядом с другой строкой, но не включать эту строку в результат поиска/ 

<https://regex101.com/r/a2S3Z3/1> 


https://regex101.com/r/9oiWUZ/1

In [66]:
import re

regex = r"^ *\d+"

test_str = ("50 Cactus Freeway, Phoenix, AZ\n"
	"777 5th Ave, New York, NY\n"
	"50 Market Blvd, Oakland, CA\n"
	"     505 Nord Freeway, Phoenix, AZ\n"
	"57 New St., Oakland, CA\n")
print(test_str)

print(r'=='*20)

subst = "xxx"

# You can manually specify the number of replacements by changing the 4th argument
result = re.sub(regex, subst, test_str, 0, re.MULTILINE | re.IGNORECASE|re.UNICODE)

if result:
    print (result)

50 Cactus Freeway, Phoenix, AZ
777 5th Ave, New York, NY
50 Market Blvd, Oakland, CA
     505 Nord Freeway, Phoenix, AZ
57 New St., Oakland, CA

xxx Cactus Freeway, Phoenix, AZ
xxx 5th Ave, New York, NY
xxx Market Blvd, Oakland, CA
xxx Nord Freeway, Phoenix, AZ
xxx New St., Oakland, CA



In [69]:
import re

regex = re.compile( r"^ *\d+",re.MULTILINE | re.IGNORECASE|re.UNICODE)



test_str = ("50 Cactus Freeway, Phoenix, AZ\n"
	"777 5th Ave, New York, NY\n"
	"50 Market Blvd, Oakland, CA\n"
	"     505 Nord Freeway, Phoenix, AZ\n"
	"57 New St., Oakland, CA\n")
print(test_str)

print(r'=='*20)

subst = "xxx"

# You can manually specify the number of replacements by changing the 4th argument
result = re.sub(regex, subst, test_str)

if result:
    print (result)

50 Cactus Freeway, Phoenix, AZ
777 5th Ave, New York, NY
50 Market Blvd, Oakland, CA
     505 Nord Freeway, Phoenix, AZ
57 New St., Oakland, CA

xxx Cactus Freeway, Phoenix, AZ
xxx 5th Ave, New York, NY
xxx Market Blvd, Oakland, CA
xxx Nord Freeway, Phoenix, AZ
xxx New St., Oakland, CA



In [70]:
import re

regex = re.compile( r"""^     # начало
                    \ *       # пустая строка или пробелы
                    \d+       # цепочка цифр
                    """,re.MULTILINE | re.IGNORECASE|re.UNICODE|re.VERBOSE)



test_str = ("50 Cactus Freeway, Phoenix, AZ\n"
	"777 5th Ave, New York, NY\n"
	"50 Market Blvd, Oakland, CA\n"
	"     505 Nord Freeway, Phoenix, AZ\n"
	"57 New St., Oakland, CA\n")
print(test_str)

print(r'=='*20)

subst = "xxx"

# You can manually specify the number of replacements by changing the 4th argument
result = re.sub(regex, subst, test_str)

if result:
    print (result)

50 Cactus Freeway, Phoenix, AZ
777 5th Ave, New York, NY
50 Market Blvd, Oakland, CA
     505 Nord Freeway, Phoenix, AZ
57 New St., Oakland, CA

xxx Cactus Freeway, Phoenix, AZ
xxx 5th Ave, New York, NY
xxx Market Blvd, Oakland, CA
xxx Nord Freeway, Phoenix, AZ
xxx New St., Oakland, CA



In [11]:
import re

regex = r"^\d+\s+"

50 Cactus Freeway, Phoenix, AZ
777 5th Ave, New York, NY
50 Market Blvd, Oakland, CA
505 Nord Freeway, Phoenix, AZ
57 New St., Oakland, CA
Cactus Freeway, Phoenix, AZ
5th Ave, New York, NY
Market Blvd, Oakland, CA
Nord Freeway, Phoenix, AZ
New St., Oakland, CA


In [36]:
import re
## re.VERBOSE == re.X   ## игнорировать пробелы кроме \пробел
                        ## игнорировать переходы на новую строку 
                        ## игнорироватькоментарии

# regex = re.compile(r"^  \d+   \s+",re.MULTILINE |re.UNICODE | re.IGNORECASE|re.VERBOSE)
regex = re.compile(r"""^       # якорь начала строки 
                        \d+    # любая цифра, повторение 1 или более
                        \s+    #                """,
                   re.M |re.U | re.I|re.X)



print(type(regex))

test_str = ("50 Cactus Freeway, Phoenix, AZ\n"
	"777 5th Ave, New York, NY\n"
	"50 Market Blvd, Oakland, CA\n"
	"505 Nord Freeway, Phoenix, AZ\n"
	"57 New St., Oakland, CA")

print(test_str)

print('='*40)


subst = ""

# You can manually specify the number of replacements by changing the 4th argument
result = re.sub(regex, subst, test_str )

if result:
    print (result)

<class 're.Pattern'>
50 Cactus Freeway, Phoenix, AZ
777 5th Ave, New York, NY
50 Market Blvd, Oakland, CA
505 Nord Freeway, Phoenix, AZ
57 New St., Oakland, CA
Cactus Freeway, Phoenix, AZ
5th Ave, New York, NY
Market Blvd, Oakland, CA
Nord Freeway, Phoenix, AZ
New St., Oakland, CA


In [72]:
import re

regex = r"^ *(\d+) (.*?), *(.*), *([A-Z]{2})$"

test_str = ("50 Cactus Freeway, Phoenix, AZ\n"
	"777 5th Ave, New York, NY\n"
	"50 Market Blvd, Oakland, CA\n"
	"     505 Nord Freeway, Phoenix, AZ\n"
	"57 New St., Oakland, CA\n"
	"eeeeeeeee\n"
	"Папа у Васи силен в математике\n"
	"Илюша\n"
	"Илюшка\n"
	"Илюшечка\n"
	"ИлюшНННа\n"
	"c:/Program Files/Microsoft/Microsort Office/winword.exe\n")

matches = re.finditer(regex, test_str, re.MULTILINE | re.IGNORECASE)

for matchNum, match in enumerate(matches, start=1):
    
    print ("Match {matchNum} was found at {start}-{end}: {match}".format(matchNum = matchNum, start = match.start(), end = match.end(), match = match.group()))
    
    for groupNum in range(0, len(match.groups())):
        groupNum = groupNum + 1
        
        print ("Group {groupNum} found at {start}-{end}: {group}".format(groupNum = groupNum, start = match.start(groupNum), end = match.end(groupNum), group = match.group(groupNum)))


Match 1 was found at 0-30: 50 Cactus Freeway, Phoenix, AZ
Group 1 found at 0-2: 50
Group 2 found at 3-17: Cactus Freeway
Group 3 found at 19-26: Phoenix
Group 4 found at 28-30: AZ
Match 2 was found at 31-56: 777 5th Ave, New York, NY
Group 1 found at 31-34: 777
Group 2 found at 35-42: 5th Ave
Group 3 found at 44-52: New York
Group 4 found at 54-56: NY
Match 3 was found at 57-84: 50 Market Blvd, Oakland, CA
Group 1 found at 57-59: 50
Group 2 found at 60-71: Market Blvd
Group 3 found at 73-80: Oakland
Group 4 found at 82-84: CA
Match 4 was found at 85-119:      505 Nord Freeway, Phoenix, AZ
Group 1 found at 90-93: 505
Group 2 found at 94-106: Nord Freeway
Group 3 found at 108-115: Phoenix
Group 4 found at 117-119: AZ
Match 5 was found at 120-143: 57 New St., Oakland, CA
Group 1 found at 120-122: 57
Group 2 found at 123-130: New St.
Group 3 found at 132-139: Oakland
Group 4 found at 141-143: CA


In [85]:
import re

regex = r"^ *(\d+) (.*?), *(.*), *([A-Z]{2})$"

test_str = ("50 Cactus Freeway, Phoenix, AZ\n"
	"777 5th Ave, New York, NY\n"
	"50 Market Blvd, Oakland, CA\n"
	"     505 Nord Freeway, Phoenix, AZ\n"
	"57 New St., Oakland, CA\n"
	"eeeeeeeee\n"
	"Папа у Васи силен в математике\n"
	"Илюша\n"
	"Илюшка\n"
	"Илюшечка\n"
	"ИлюшНННа\n"
	"c:/Program Files/Microsoft/Microsort Office/winword.exe\n")


for line in test_str.splitlines():
    m_o= re.match(regex,line,re.I|re.M)
    if m_o :
        print(m_o[4],m_o[3])
    

AZ Phoenix
NY New York
CA Oakland
AZ Phoenix
CA Oakland


In [29]:
import re

regex = re.compile( (r"^  " # начало
                     r"\d+ " # цепочка цифр 1 или более длинной
                     r"\s+"  # цепочка пробельных символов 1 или более длинной
                     ),   
                     re.MULTILINE |re.UNICODE | re.IGNORECASE|re.VERBOSE)

print(type(regex))

test_str = ("50 Cactus Freeway, Phoenix, AZ\n"
	"777 5th Ave, New York, NY\n"
	"50 Market Blvd, Oakland, CA\n"
	"505 Nord Freeway, Phoenix, AZ\n"
	"57 New St., Oakland, CA")

print(test_str)

print('='*40)


subst = ""

# You can manually specify the number of replacements by changing the 4th argument
result = re.sub(regex, subst, test_str )

if result:
    print (result)

<class 're.Pattern'>
50 Cactus Freeway, Phoenix, AZ
777 5th Ave, New York, NY
50 Market Blvd, Oakland, CA
505 Nord Freeway, Phoenix, AZ
57 New St., Oakland, CA
Cactus Freeway, Phoenix, AZ
5th Ave, New York, NY
Market Blvd, Oakland, CA
Nord Freeway, Phoenix, AZ
New St., Oakland, CA


* * *

### Примеры применения регулярных выражений в pandas

In [1]:
%%writefile custdata.txt
cust_id; name; address; phone; cust_type
100; The Camera Store; 300 Broad St., New York, NY; 2125552222; Retail
101; Wilson Security; 50 Cactus Freeway, Phoenix, AZ; 6095550000; End User
102; City Copy Shop; 777 5th Ave, New York, NY; 2125551111; End User
103; Computer Planet; 100 Boylston St., Boston, MA; 6175553333; Retail
104; Hi Tech Supply; 50 Market Blvd, Oakland, CA; 5103334444; Distributo

Writing custdata.txt


In [2]:
%%writefile addresses.txt
300 Broad St., New York, NY
50 Cactus Freeway, Phoenix, AZ
777 5th Ave, New York, NY
100 Boylston St., Boston, MA
50 Market Blvd, Oakland, CA
505 Nord Freeway, Phoenix, AZ
57 New St., Oakland, CA
731 Fondren Drv, Houston, TX
638 Voss St., Houston, TX
291 Berry St., Bellaire, TX
975 Fire Oak Ave, Humble, TX
980 Dallas St., Houston, TX
450 Stone Ave, Houston, TX

Writing addresses.txt


In [37]:
import pandas as pd

In [41]:
cust0=pd.read_csv("custdata.txt", sep=';',header=0 )
cust0.keys() # В имена полей "втесались" пробелы

Index(['cust_id', ' name', ' address', ' phone', ' cust_type'], dtype='object')

* * *

Укажем `sep=';\s*'`, чтобы разделение было по `;` с необязательной цепочкой пробелов `\s*`

In [42]:
cust=pd.read_csv("custdata.txt", sep=';\s*',engine="python",header=0 )

In [40]:
cust.keys()

Index(['cust_id', 'name', 'address', 'phone', 'cust_type'], dtype='object')

In [10]:
cust["name"]

0    The Camera Store
1     Wilson Security
2      City Copy Shop
3     Computer Planet
4      Hi Tech Supply
Name: name, dtype: object

In [11]:
bool_index=cust["name"].str.contains('Store|Shop') # re  альтернатива |
bool_index

0     True
1    False
2     True
3    False
4    False
Name: name, dtype: bool

In [12]:
cust.loc[bool_index, ["name", "cust_type"]]

name cust_type
0  The Camera Store    Retail
2    City Copy Shop  End User

* * *

Заменяем номера домов  в начале адресов на строку `xxx`

In [44]:
cust.address.str.replace(r"^\d+","xxx")

0        xxx Broad St., New York, NY
1    xxx Cactus Freeway, Phoenix, AZ
2          xxx 5th Ave, New York, NY
3       xxx Boylston St., Boston, MA
4       xxx Market Blvd, Oakland, CA
Name: address, dtype: object

In [45]:
cust.address.str.replace(r"^(\d+)(.*?)," , r"\2 dom \1") # re groups and \1 \2

0        Broad St. dom 300 New York, NY
1     Cactus Freeway dom 50 Phoenix, AZ
2          5th Ave dom 777 New York, NY
3       Boylston St. dom 100 Boston, MA
4        Market Blvd dom 50 Oakland, CA
Name: address, dtype: object

## Использование модуля re

In [40]:
import re
## import regex

Основные методы в модуле `re`

```python
re.match(pattern, text)  # match -- сопоставлять, подходить
    # --> None или match-object. Сопоставляет pattern с началом строки text
re.search(pattern, text) 
    # --> None или match-object. Сопоставляет pattern с любой частью строки text
re.findall(pattern, text)
    # --> список всех найденых подстрок 
re.split(pattern, text) 
    # --> список строк, на которые разбивается текст разделителями, подходищими под шаблон pattern
re.sub(pattern, replacement,text) # sub - substitute (подставлять, заменять)
    # --> новая версия text, где подстроки, соответствующие pattern заменены на replacement 
```

In [44]:
spisok_str="1345,23434,2343,  3434,\t3434"

In [45]:
spisok_str.split(",")

['1345', '23434', '2343', '  3434', '\t3434']

In [47]:
re.split(r',\s*',spisok_str)

['1345', '23434', '2343', '3434', '3434']

In [48]:
spisok_str="1345,23434;2343,  3434,\t3434"
re.split(r'[,;]\s*',spisok_str)

['1345', '23434', '2343', '3434', '3434']

In [50]:
re.findall(r'\d+',spisok_str)

['1345', '23434', '2343', '3434', '3434']

In [51]:
spisok_str="134.5,23434;23.43,  3434,\t3434"
re.findall(r'[.\d]+',spisok_str)

['134.5', '23434', '23.43', '3434', '3434']

In [57]:
answer='Yes'
m=re.match('(yes|y|д|да)$',answer,re.IGNORECASE)

In [58]:
if m :
    print("пользователь согласен")
    print(m)
else:
    print(type(m))
    print(answer)

пользователь согласен
<re.Match object; span=(0, 3), match='Yes'>


### Пример. Поиск имен в модуле по шаблону

Часто в модуле много функций и констант, а у объекта много методов. И тогда  в длинном списке, выдаваемом `dir(obj)` мы часто хотим найти имя, которое помним неточно. Регулярные выражения нам в помощь 

Найдем имена в которых есть строка `_compile` сначала в начале строки (`match()`), а потом а в любом месте строки (`search()`)

In [16]:
import re
re_ids = dir(re) # все имена, определенные в модуле re

In [17]:
print(*re_ids)

A ASCII DEBUG DOTALL I IGNORECASE L LOCALE M MULTILINE Match Pattern RegexFlag S Scanner T TEMPLATE U UNICODE VERBOSE X _MAXCACHE __all__ __builtins__ __cached__ __doc__ __file__ __loader__ __name__ __package__ __spec__ __version__ _cache _compile _compile_repl _expand _locale _pickle _special_chars_map _subx compile copyreg enum error escape findall finditer fullmatch functools match purge search split sre_compile sre_parse sub subn template


In [18]:
match_obj1 = re.match("A", 'ASCII')
match_obj2 = re.match("A", 'LOCALE')
match_obj3 = re.search("A", 'LOCALE')

In [19]:
print(match_obj1, match_obj2, match_obj3,sep='\n')

<re.Match object; span=(0, 1), match='A'>
None
<re.Match object; span=(3, 4), match='A'>


In [20]:
for s in re_ids:
    if re.match(pattern="_compile",string=s):
        print(s)

_compile
_compile_repl


In [23]:
for s in re_ids:
    if re.search(pattern="_compile",string=s):
        print(s)

_compile
_compile_repl
sre_compile


то же самое без цикла, при помощи списковых включений

In [21]:
[ s for s in re_ids if re.match(pattern="_compile",string=s) ]

['_compile', '_compile_repl']

In [22]:
[ s for s in re_ids if re.search(pattern="_compile",string=s) ]

['_compile', '_compile_repl', 'sre_compile']

In [24]:
 ssss= "one,two;three.four.five six"

обычная строковая фукнция `split()`  не может использовать продвинутые шаблоны, а ищет конкретные символы

In [25]:
ssss.split(".")

['one,two;three', 'four', 'five six']

In [109]:
# re.split?

In [26]:
re.split("[,;. ]",ssss)

['one', 'two', 'three', 'four', 'five', 'six']

В модуле `re` не поддерживаются классы символов внутри квадратных скобок и нужно писать явные списки и диапазоны символов или пользоваться Perl-Compatible-RE 

In [112]:
re.split("[[:punct:][:blank:]]",ssss) # строка не разбилась на части

['one,two;three.four.five six']

In [27]:
re.split(r"\W",ssss) # делим по символам, которые не символы слова

['one', 'two', 'three', 'four', 'five', 'six']

## Отступление от темы RE про строковые литералы

строковые константы с тройными кавычками/апострофами могут содержать **многострочный текст**


```python
'''многострочный 
текст'''

""" многострочный 
текст"""
```

В обычных строках многие сочетания (escape-последовательности) типа \n, \t интепретируются как один специальный символ (new line, tab). И для того, чтобы избежать специальной интерпретации символа "\" его либо удваивают "\\" либо используют сырые **raw** строковые константы. 

Сырые **raw** строковые константы: 

```python
r"C:\Users\Student\Documents"
```

Чтобы не получать трудно диагностируемых проблем при задании шаблонов, используйте сырые строковые литералы для задания регулярных выражений.

```python
re.search(r"\d+\t\d+, text )
```

Многострочные литералы:

In [93]:
mline="""'1st line
nd line
3rd line'"""
mline

"'1st line\nnd line\n3rd line'"

In [95]:
print(mline)

'1st line
nd line
3rd line'


Удвоение символа backslash `\\` чтобы изобразить один `\`

In [98]:
print("\\n") # 2 символа: \ и n 

\n


In [19]:
print("C:\\users\\student\\documents")

C:\users\student\documents


### raw stings 

Сырые строковые литералы `r" "` с префиксом `r`  полезны при написании путей файловой Windows и при написании регулярных выражений

In [27]:
r"\n"

'\\n'

In [28]:
print(r"\n")

\n


In [21]:
print(r"C:\users\student\documents")

C:\users\student\documents


А так сработает функция `repr()` которая приводит распечатываемое выражение к синтаксису, уместному в программе:

In [22]:
r"C:\users\student\documents"

'C:\\users\\student\\documents'

In [115]:
file=open("addresses.txt")

In [116]:
type(file)

_io.TextIOWrapper

In [117]:
for line in file:
    print(line,end="") # там строки были и так с "\n" на конце

300 Broad St., New York, NY
50 Cactus Freeway, Phoenix, AZ
777 5th Ave, New York, NY
100 Boylston St., Boston, MA
50 Market Blvd, Oakland, CA
505 Nord Freeway, Phoenix, AZ
57 New St., Oakland, CA
731 Fondren Drv, Houston, TX
638 Voss St., Houston, TX
291 Berry St., Bellaire, TX
975 Fire Oak Ave, Humble, TX
980 Dallas St., Houston, TX
450 Stone Ave, Houston, TX


In [118]:
addr_list=file.readlines() # файл и так был дочитан до конца

In [119]:
addr_list

[]

In [120]:
file.close()

In [125]:
file=open("addresses.txt")
addr_list= [ s.rstrip() for s in file ]
file.close()

In [126]:
addr_list

['300 Broad St., New York, NY',
 '50 Cactus Freeway, Phoenix, AZ',
 '777 5th Ave, New York, NY',
 '100 Boylston St., Boston, MA',
 '50 Market Blvd, Oakland, CA',
 '505 Nord Freeway, Phoenix, AZ',
 '57 New St., Oakland, CA',
 '731 Fondren Drv, Houston, TX',
 '638 Voss St., Houston, TX',
 '291 Berry St., Bellaire, TX',
 '975 Fire Oak Ave, Humble, TX',
 '980 Dallas St., Houston, TX',
 '450 Stone Ave, Houston, TX']

In [127]:
address = addr_list[2]

In [128]:
match_obj = re.match(r"\d+",address)

Объект сопоставления *match object*  содержит и границы индексов участка, с которым совпал шаблон (span)  и найденную строку (match)

In [129]:
match_obj

<re.Match object; span=(0, 3), match='777'>

In [130]:
address[0:3]

'777'

In [131]:
match_obj

<re.Match object; span=(0, 3), match='777'>

match_obj.start()

In [132]:
match_obj.end()

3

In [133]:
address[match_obj.start():match_obj.end()]

'777'

Проще получить найденную подстроку при помощи `match_obj.group(0)`

In [134]:
match_obj.group(0)

'777'

теперь ищем не в начале строки

In [135]:
mo2 = re.search(r"\s\w+",address)

In [136]:
mo2.group(0)

' 5th'

In [28]:
mo=re.match(r"(\d+) (.*?), (.*?), (\w{2})$", '975 Fire Oak, Humble, TX' )

In [29]:
mo.group(0) # весь шаблон соответсвует группе 0 

'975 Fire Oak, Humble, TX'

In [30]:
[mo.group(1), mo.group(2), mo.group(3), mo.group(4)]

['975', 'Fire Oak', 'Humble', 'TX']

In [31]:
mo.groups()

('975', 'Fire Oak', 'Humble', 'TX')

## Нежадные шаблоны

В предыдущем примере использованы *нежадные шаблоны*.  По умолчанию шаблоны типа `".*,"` жадные --- они сопоставляются с подстрокой максимальной длинны из всех подходящих для шаблона подстрок (до последней запятой). А варианты суффиксов повторения с вопросом `*?, +?, ??, {1,3}?` -- не жадные и сопостовляются с подстрокой минимальной длинны. `.*,` дойдет до первой запятой.  

## Комментирование частей RE

Дополнительный параметр flag со значением `re.X` или  `re.VERBOSE` позволяет добавлять в шаблон игнорируемые  пробелы, переходы на новую строку и комментарии со знаком # для того чтобы сделать сложные шаблоны понятнее. Предварённые backslash "\\" пробельные символы не игнорируются.

In [46]:
mo=re.match(r"""
            (\d+)\  # group(1): номер дома; цифры до первого пробела
            (.*?),\  # group(2): Название улицы до 1й запятой. Нежадный шаблон.
            (.*?),\  # group(3): город
            ([A-Z]{2})$ # group(4): перед концом строки штат, записанный 2мя заглавными буквами""", 
            '975 Fire Oak, Humble, TX', 
            re.VERBOSE )

In [142]:
mo.group(3)

'Humble'

## Именнованные группы

```
(?P<имя_группы>шаблон)
```

In [82]:
mo=re.match(r"""
            (?P<дом>\d+)\  # group("дом"): номер дома; цифры до первого пробела
            (?P<улица>.*?),\  # group("улица"): Название улицы после номера дома и до 1й запятой. Нежадный шаблон.
            (?P<город>.*?),\  # group("город"): город
            (?P<штат>[A-Z]{2})$ # group("штат"): перед концом строки штат, записанный 2мя заглавными буквами""", 
            '975 Fire Oak, Humble, TX', 
            re.X )

In [83]:
mo.group("город")

'Humble'

In [145]:
mo[1] # Группы 0 и 1-4 могут быть извлечены из match-object через индексы 0 и 1..4

'975'

In [34]:
mo.groupdict()

{'дом': '975', 'улица': 'Fire Oak', 'город': 'Humble', 'штат': 'TX'}

In [147]:
mo["улица"]

'Fire Oak'

* * *


# Упражнения к части 2

## Упражнение 1

Выдайте список имен, описанных в модуле  re, в которых есть слово loc. Установите дополнительный     параметр `flag` чтобы регистр символов при сопоставлении игнорировать 

предполагаемый  результат:
```
['LOCALE', '_locale']
```

In [148]:
[ s for s in  dir(re) if re.search("loc",s,re.I)]

['LOCALE', '_locale']

## Упражнение 2 
Переменной `addr_file_fragment` присвоен текст с  адресами email и баллами студентов

In [149]:
addr_file_fragment = """full_email;score
Sara Evans <sarae428@yahoo.com>;97
Sara Parks <smparks22@yahoo.com>;116
Sara Stone <sarastone@walsh.edu>;60
Sarah Barr <sarahl.barr@gmail.com>;109
Sarah Boron <boron120@yahoo@com>;83
Sarah Buell <Buellsarahc@yahoo.com>;79
Sarah Lucas <sarah@buffingtonagency.com>;64
Sarah Shafer <wall_eye49@yahoo.com>;71
Sarah Tombin <jess_192001@yahoo.com>;74
Scott Archer <dscottarch@yahoo.com>;33
Sean Shirley <sean.shirley@emerson.com>;109
Shanda Hudson <semhudson@gmail.com>;65
Shannon West <shananagines@yahoo.com>;120
Shari Turrin <shari2573@yahoo.com>;77
"""

Выведите пронумерованные строки, где в почтоовм адресе ошибка ( 2 знака @ ) 

Решение:


преполагаемый результат:


```
6:	Sarah Boron <boron120@yahoo@com>;83
```

In [150]:
nline=1
for s in addr_file_fragment.split("\n"):
    if  re.search(r"@.*@",s):
        print("{}:\t{}".format(nline,s))
    nline+=1

6:	Sarah Boron <boron120@yahoo@com>;83


* * *

## Упражнение 3

Переменной `addr_file_fragment` из упражения 2 присвоен текст с  адресами email и баллами студентов.

сделайте шаблон, делящий строку исходных данных на такие именнованные группы:

|  описание шаблона         | имя группы |
|:--------------------------|------------------------------------:|
| цепочка любых символов после которой пробел | fname |  
| цепочка непробелов после которой пробел | lname |
| цепочка любых символов внутри знаков < и > | email |
| цепочка цифр поле символа ; | score |
                  

Выведите эту информацию в таком format'е  : 
```python
"First name: {fname:>9} | Last name:{lname:>9} | email:{email:>27}" 
```

а для строк исходных данных, не подходящих под шаблон, чтобы выдавалось 79 знаков "-"
и получилась такая таблица:

```
-------------------------------------------------------------------------------
First name:      Sara | Last name:    Evans | email:         sarae428@yahoo.com
First name:      Sara | Last name:    Parks | email:        smparks22@yahoo.com
First name:      Sara | Last name:    Stone | email:        sarastone@walsh.edu
First name:     Sarah | Last name:     Barr | email:      sarahl.barr@gmail.com
First name:     Sarah | Last name:    Boron | email:         boron120@yahoo@com
First name:     Sarah | Last name:    Buell | email:      Buellsarahc@yahoo.com
First name:     Sarah | Last name:    Lucas | email: sarah@buffingtonagency.com
First name:     Sarah | Last name:   Shafer | email:       wall_eye49@yahoo.com
First name:     Sarah | Last name:   Tombin | email:      jess_192001@yahoo.com
First name:     Scott | Last name:   Archer | email:       dscottarch@yahoo.com
First name:      Sean | Last name:  Shirley | email:   sean.shirley@emerson.com
First name:    Shanda | Last name:   Hudson | email:        semhudson@gmail.com
First name:   Shannon | Last name:     West | email:     shananagines@yahoo.com
First name:     Shari | Last name:   Turrin | email:        shari2573@yahoo.com
-------------------------------------------------------------------------------
```

## Упражение 4

Напишите вызов функции sub() чтобы оно выдавала строку "Гав! Гав Гав!" Предварительно заменив все вхождения строки "Гав" на "Мяу"

предполагаемый результат

'Мяу! Мяу Мяу!'

In [151]:
re.sub("Гав", "Мяу", "Гав! Гав Гав!")

'Мяу! Мяу Мяу!'

* * * 

## Упражнение 5

В файле `bad_bad_addresses.dos.csv` 1177 строк.  

путь к файлу `"https://github.com/bivni/tidy_demo/raw/master/data/bad_addresses.dos.csv"`

В первой строке заголовок с  названиями столбцов:
```
full_email;score
Sara Evans <sarae428@yahoo.com>;97
Sara Parks <smparks22@yahoo.com>;116
Sara Stone <sarastone@walsh.edu>;60
Sarah Barr <sarahl.barr@gmail.com>;109
Sarah Boron <boron120@yahoo@com>;83
Sarah Buell <Buellsarahc@yahoo.com>;79
Sarah Lucas <sarah@buffingtonagency.com>;64
Sarah Shafer <wall_eye49@yahoo.com>;71
Sarah Tombin <jess_192001@yahoo.com>;74
Scott Archer <dscottarch@yahoo.com>;33
Sean Shirley <sean.shirley@emerson.com>;109
Shanda Hudson <semhudson@gmail.com>;65
Shannon West <shananagines@yahoo.com>;120
Shari Turrin <shari2573@yahoo.com>;77
```

Найдите в Internet готовый шаблон для корректных email адресов  и найдите 2 строки с некоректными email. 



In [157]:
pat=re.compile(r"""^(?P<fname>.*)\      # group(fname)  
                    (?P<lname>\S+)\     # group(lname)
                    <(?P<email>.+)>; # group(email) 
                    (?P<score>\d+)$ # group(score)
                """, re.X)
output_format_str = "First name: {fname:>9} | Last name:{lname:>9} | email:{email:>27}"

for s in addr_file_fragment.split("\n"):
    mo=pat.match(s)
    if mo:
        print(output_format_str.format(**mo.groupdict()))
    else:
        print(79*"-")

-------------------------------------------------------------------------------
First name:      Sara | Last name:    Evans | email:         sarae428@yahoo.com
First name:      Sara | Last name:    Parks | email:        smparks22@yahoo.com
First name:      Sara | Last name:    Stone | email:        sarastone@walsh.edu
First name:     Sarah | Last name:     Barr | email:      sarahl.barr@gmail.com
First name:     Sarah | Last name:    Boron | email:         boron120@yahoo@com
First name:     Sarah | Last name:    Buell | email:      Buellsarahc@yahoo.com
First name:     Sarah | Last name:    Lucas | email: sarah@buffingtonagency.com
First name:     Sarah | Last name:   Shafer | email:       wall_eye49@yahoo.com
First name:     Sarah | Last name:   Tombin | email:      jess_192001@yahoo.com
First name:     Scott | Last name:   Archer | email:       dscottarch@yahoo.com
First name:      Sean | Last name:  Shirley | email:   sean.shirley@emerson.com
First name:    Shanda | Last name:   Hud

## Упражнение 6

импортируйте при помощи библиотеки pandas файл из предыдущего упражнения в DataFrame.

In [35]:
import numpy as np
import pandas as pd
df = pd.read_csv("https://github.com/bivni/tidy_demo/raw/master/data/bad_addresses.dos.csv",sep=";"
                 ,dtype={"full_email":str, "score":np.int32}
                )
df.tail(6)

full_address  score
1170  William A Tolley <billthistoo@sbcglobal.com>     25
1171               William Mossor <bdm523@aol.com>     51
1172         Yvette Green <yvette.green@yahoo.com>     99
1173          Yvette March <yvettemarch@yahoo.com>     55
1174            Yvette Marman <66ymm99@sssnet.com>     30
1175                Zandy Rote <zrote@hotmail.com>    109

Используя метод `df.full_address.str.extract(шаблон_с_группами)` получите DataFrame `contacts` с отдельными столбцами  с названиями групп шаблона `fname`, `midinit`, `lname`, `email` куда распределите данные из стобца `full_email`. Выдайте последние 10 строк этого DataFrame `contacts`

Предполагаемый ответ: 

| NN   | fname     | lname    | email                      |
|------|-----------|----------|----------------------------|
| 1166 | Walt      | Cerrato  | Wc32124@gmail.com          |
| 1167 | Wendy     | Kelly    | danandwendy@mac.com        |
| 1168 | Wendy     | Menegay  | wendymenegay18@hotmail.com |
| 1169 | Wilda     | Phillips | sales@phillipsstamping.com |
| 1170 | William A | Tolley   | billthistoo@sbcglobal.com  |
| 1171 | William   | Mossor   | bdm523@aol.com             |
| 1172 | Yvette    | Green    | yvette.green@yahoo.com     |
| 1173 | Yvette    | March    | yvettemarch@yahoo.com      |
| 1174 | Yvette    | Marman   | 66ymm99@sssnet.com         |
| 1175 | Zandy     | Rote     | zrote@hotmail.com          |

In [36]:
mail_pattern=re.compile(
               r"""^(?P<fname>.*)\s     # group(fname)
                    (?P<lname>\S+)\s   # group(lname)
                    <(?P<email>.+)>$ # group(score)
                """, re.X)

contacts=df.full_address.str.extract(mail_pattern, expand=True)

contacts.tail(10)

fname     lname                       email
1166       Walt   Cerrato           Wc32124@gmail.com
1167      Wendy     Kelly         danandwendy@mac.com
1168      Wendy   Menegay  wendymenegay18@hotmail.com
1169      Wilda  Phillips  sales@phillipsstamping.com
1170  William A    Tolley   billthistoo@sbcglobal.com
1171    William    Mossor              bdm523@aol.com
1172     Yvette     Green      yvette.green@yahoo.com
1173     Yvette     March       yvettemarch@yahoo.com
1174     Yvette    Marman          66ymm99@sssnet.com
1175      Zandy      Rote           zrote@hotmail.com